atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java

```diff
@@ -84,18 +84,33 @@ public HikariCPConnectionManager(ConnectionConfig connConfig) {
 
     @Override
     public Connection getConnection() throws SQLException {
-        long start = System.currentTimeMillis();
-        try {
-            return getConnectionInternal();
-        } finally {
-            long now = System.currentTimeMillis();
-            long elapsed = now - start;
-            if (elapsed > 1000) {
-                log.warn("Waited {}ms for connection", elapsed);
-                logPoolStats();
-            } else {
-                log.debug("Waited {}ms for connection", elapsed);
+        while (true) {
+            long start = System.currentTimeMillis();
+            HikariDataSource dataSourcePool;
+            Connection conn;
+            try {
+                dataSourcePool = getDataSourcePool();
+                conn = dataSourcePool.getConnection();
+            } finally {
+                long now = System.currentTimeMillis();
+                long elapsed = now - start;
+                if (elapsed > 1000) {
+                    log.warn("Waited {}ms for connection", elapsed);
+                    logPoolStats();
+                } else {
+                    log.debug("Waited {}ms for connection", elapsed);
+                }
+            }
+
+            try {
+                testConnection(conn);
+            } catch (SQLException e) {
+                log.error("Dropping connection which failed validation", e);
+                dataSourcePool.evictConnection(conn);
+                continue;
```

Comment:
we need to test the situation where this could infinite loop, may want to only retry some N number of times and throw if we can't get a valid connection